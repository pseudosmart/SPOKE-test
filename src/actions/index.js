export * from './basket'
export * from './products'