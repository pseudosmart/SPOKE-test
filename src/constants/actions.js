// BASKET
export const ADD_TO_BASKET = 'ADD_TO_BASKET'
export const REMOVE_FROM_BASKET = 'REMOVE_FROM_BASKET'
export const UPDATE_TOTAL = 'UPDATE_TOTAL'

// PRODUCTS

export const REQUEST_PRODUCTS = 'REQUEST_PRODUCTS'
export const RECEIVE_PRODUCTS = 'RECEIVE_PRODUCTS'