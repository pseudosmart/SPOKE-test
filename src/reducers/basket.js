import {
  ADD_TO_BASKET,
  REMOVE_FROM_BASKET, //TODO
  UPDATE_TOTAL
} from '../constants/actions'

const intialState = {
  currentBasket: [],
  total: 0
}

export const getCurrentTotal = (currentState) => {
  return currentState.currentBasket.reduce((total, basketItem) => {
    return parseInt(basketItem.price, 10) + total
  }, 0)
}

export default function (currentState = intialState, action) {
  switch (action.type) {
    case ADD_TO_BASKET:
      return {
        ...currentState,
        currentBasket: [...currentState.currentBasket, action.payload]
      }
      case UPDATE_TOTAL:
      return {
        ...currentState,
        total: getCurrentTotal(currentState).toFixed(2)
      }

    default: 
    return currentState
    }
}