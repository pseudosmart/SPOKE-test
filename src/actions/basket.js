import { ADD_TO_BASKET, REMOVE_FROM_BASKET, UPDATE_TOTAL } from '../constants/actions'

export const addToBasket = product => dispatch => {
  dispatch({
    type: ADD_TO_BASKET,
    payload: product })
  dispatch({
    type: UPDATE_TOTAL
  })
}

export const removeFromBasket = product => ({
  type: REMOVE_FROM_BASKET,
  payload: product
})



