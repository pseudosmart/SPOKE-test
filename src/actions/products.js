import { REQUEST_PRODUCTS, RECEIVE_PRODUCTS } from '../constants/actions'
import { graphQL } from '../helpers'

export const fetchProducts = () => dispatch => {
  dispatch({type: REQUEST_PRODUCTS})
  return graphQL.query().then(products => {
    return dispatch(productFetchSuccess(products))
  })
}

export const productFetchSuccess = products => ({
  type: RECEIVE_PRODUCTS,
  payload: products
})

export const productFetchError = error => ({ //TODO
  type: RECEIVE_PRODUCTS,
  error: true,
  payload: error
})
