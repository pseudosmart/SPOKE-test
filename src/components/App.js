import React, {Component} from 'react'
import { connect } from 'react-redux'
import { fetchProducts, addToBasket } from '../actions'

import ProductTile from './ProductTile'
import Basket from './Basket'

export class App extends Component {
  constructor(props) {
    super(props)
    this.props.fetchProducts()
  }


  render () {
    const { products, addToBasket } = this.props

    if (!products) return <h1>Getting Data...</h1> //Would look at loading animation using the fetching property from the product store

    return (
      <div className='row'>
        <div className='col-12 col-lg-2 order-lg-12'>
          <Basket />
        </div>

        <div className='col-12 col-lg-10'>
        { products.map(product => ( 
          <ProductTile key={product.id} product={product} basketHandler={addToBasket} />
        ))}
        </div>


      </div>
    )
  }
}


export const mapDispatchToProps = (dispatch) => ({
  fetchProducts: () => dispatch(fetchProducts()),
  addToBasket: (product) => dispatch(addToBasket(product))
})

export const mapStateToProps = (state) => {
  return {
    products: state.products.products, //horrible named nesting
    basket: state.basket
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App)


