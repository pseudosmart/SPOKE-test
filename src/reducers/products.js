import { REQUEST_PRODUCTS, RECEIVE_PRODUCTS } from '../constants/actions'

const flattenProducts = products => {
  const flatProducts = products.data.products.edges.reduce((productList, product) => { // Would need much better error handling
    const {id, images, productType, title, tags, priceRange, updatedAt} = product.node
    return [
      ...productList,
      {
        id,
        image: images.edges.length ? images.edges.pop().node.originalSrc : '',
        productType,
        title,
        tags,
        price: parseInt(priceRange.maxVariantPrice.amount, 10).toFixed(2),
        updatedAt
      }
    ]
  }, [])

  return flatProducts
}

export default (currentState = {}, action) => {
  switch (action.type) {
    case REQUEST_PRODUCTS:
      return {...currentState, fetching: true}

    case RECEIVE_PRODUCTS:
      const formattedProducts = flattenProducts(action.payload)
      return { ...currentState, products: formattedProducts, fetching: false} //TODO use the fetching
      
      
      
    default:
      return currentState
  }
}