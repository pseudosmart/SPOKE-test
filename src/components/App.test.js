import React from 'react'
import { shallow } from 'enzyme'
import { App, mapStateToProps, mapDispatchToProps} from './App'
import * as productActions from '../actions/products'
import * as basketActions from '../actions/basket'

let defaultValues = {
  products: {
    products: []
  },
  basket: {}
}

let fetchProducts = jest.fn()
let addToBasket = jest.fn()

let wrapper
const props = {
  fetchProducts,
  products: [],
  addToBasket
}

const getWrapper = (overrideProps = {}) => (
  shallow(<App {...props} {...overrideProps}/>)
)

test('renders correctly without products', () => {
  expect(getWrapper()).toMatchSnapshot()
})

test('renders correctly with products', () => {
  expect(getWrapper({products: [
    {title: 'dummy title', id: 1}
  ]})).toMatchSnapshot()
})

test('calls fetchProducts', () => {
  expect(props.fetchProducts).toHaveBeenCalled()
})

test('mapStateToProps', () => {
  expect(mapStateToProps(defaultValues)).toMatchSnapshot()
})

describe('mapDispatchToProps', () => {
  let dispatch, props
  beforeEach(() => {
    dispatch = jest.fn()
    productActions.fetchProducts = jest.fn().mockReturnValue('FETCH_PRODUCTS')
    basketActions.addToBasket = jest.fn().mockReturnValue('ADD_TO_BASKET')
    props = mapDispatchToProps(dispatch)
  })

  test('it returns the correct props', () => {
    expect(props).toMatchSnapshot()
  })

  test('dispatches fetchProducts action', () => {
    props.fetchProducts()
    expect(dispatch).toBeCalledWith('FETCH_PRODUCTS')
    expect(productActions.fetchProducts).toBeCalled()
  })

  test('dispatches addToBasket action', () => {
    props.addToBasket('product')
    expect(dispatch).toBeCalledWith('ADD_TO_BASKET')
    expect(basketActions.addToBasket).toBeCalledWith('product')
  })
})