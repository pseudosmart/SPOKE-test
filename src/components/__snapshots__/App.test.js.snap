// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps it returns the correct props 1`] = `
Object {
  "addToBasket": [Function],
  "fetchProducts": [Function],
}
`;

exports[`mapStateToProps 1`] = `
Object {
  "basket": Object {},
  "products": Array [],
}
`;

exports[`renders correctly with products 1`] = `ShallowWrapper {}`;

exports[`renders correctly without products 1`] = `ShallowWrapper {}`;
